libs/langchain/ragstack_langchain/graph_store/extractors/link_extractor.py

```diff
@@ -0,0 +1,48 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from itertools import repeat
+from typing import Any, Dict, Generic, Iterable, Optional, Set, TypeVar
+
+from ragstack_knowledge_store._utils import strict_zip
+
+from ragstack_langchain.graph_store.links import Link
+
+InputT = TypeVar("InputT")
+
+METADATA_LINKS_KEY = "links"
+
+
+
+class LinkExtractor(ABC, Generic[InputT]):
+    """Interface for extracting links (incoming, outgoing, bidirectional)."""
+
+    @abstractmethod
+    def extract_one(self, input: InputT, **kwargs: Any) -> Set[Link]:
+        """Add edges from each `input` to the corresponding documents.
+
+        Args:
+            input: The input content to extract edges from.
+            **kwargs: Additional keyword arguments for the extractor.
+
+        Returns:
+            Set of links extracted from the input.
+        """
+
+    def extract_batch(self,
+                      inputs: Iterable[InputT],
+                      batch_kwargs: Optional[Iterable[Dict[str, Any]]] = None,
```

Comment:
This was a bit ugly. The problem is that extractors may need additional information about the input -- for instance, HTML Link Extractor needs to know the Base URL. The thinking here is that it can be passed in via extra arguments.

An alternative, would be to have it set `InputT = Tuple[str, str]` (basically, take a tuple of the URL and the input). That may be cleaner (only one iterable). Thoughts?

Reply:
> An alternative, would be to have it set InputT = Tuple[str, str] 

I prefer this solution. And maybe InputT could be typed stronger: InpuT = HtmlLinkInput